elk/extraction/balanced_sampler.py

```diff
@@ -0,0 +1,99 @@
+from ..utils import infer_label_column
+from ..math_util import stochastic_round_constrained
+from dataclasses import dataclass, field, InitVar
+from datasets import IterableDataset
+from itertools import cycle
+from random import Random
+from torch.utils.data import IterableDataset as TorchIterableDataset
+from typing import Iterator, Optional
+import numpy as np
+
+
+@dataclass
+class BalancedSampler(TorchIterableDataset):
+    """
+    Approximately balances a binary classification dataset in a streaming fashion.
+    Written mostly by GPT-4.
+
+    Args:
+        dataset (IterableDataset): The HuggingFace IterableDataset to balance.
+        label_col (Optional[str], optional): The name of the column containing the
+            binary label. If not provided, the label column will be inferred from
+            the dataset features. Defaults to None.
+        buffer_size (int, optional): The total buffer size to use for balancing the
+            dataset. This value should be divisible by 2, as it will be equally
+            divided between the two binary label values (0 and 1). Defaults to 1000.
+    """
+
+    dataset: IterableDataset
+    label_counts: np.ndarray = field(default_factory=lambda: np.zeros(2))
+    seed: InitVar[int] = 42
+
+    def __post_init__(self, seed: int):
+        self.rng = np.random.default_rng(seed)
+
+    def __iter__(self):
+        for sample in self.dataset:
```

Comment:
If the dataset we're streaming from isn't shuffled (e.g. all the movie reviews about batman come first) our sampling will be distributionally incorrect, no?

Reply:
Also one of the tests broke.

Reply:
Yeah working on the test thing, it's a pretty bizarre error.

The shuffling isn't really an issue. See the HF docs on this https://huggingface.co/docs/datasets/stream#shuffle. We just need to make sure we're actually calling `.shuffle`, which I thought we were, but it's entirely possible I'm just misremembering.